lib/sles4sap/sdaf_library.pm

```diff
@@ -0,0 +1,520 @@
+    # Source file and load variables
+    assert_script_run("source $temp_file");
+
+    my $login_cmd = 'while ! az login --service-principal -u ${ARM_CLIENT_ID} -p ${ARM_CLIENT_SECRET} -t ${ARM_TENANT_ID}; do sleep 10; done';
+    assert_script_run($login_cmd, timeout => 5 * 60);
+
+    my $subscription_id = script_output('az account show -o tsv --query id');
+    record_info('AZ login', "Subscription id: $subscription_id");
+
+    # Remove temp file with credentials.
+    assert_script_run("rm $temp_file");
+    return ($subscription_id);
+}
+
+=head2 create_sdaf_os_var_file
+
+    create_sdaf_os_var_file($entries);
+
+B<$entries>: ARRAYREF of entries to be appended to variable source file
+
+Creates a simple file with bash env variables and uploads it to the target host without revealing content in serial console.
+File is sourced afterwards.
+For detailed variable description check : L<https://learn.microsoft.com/en-us/azure/sap/automation/naming>
+
+=cut
+
+sub create_sdaf_os_var_file {
+    my ($entries) = @_;
+    croak 'Expected an ARRAYREF but got: ' . ref $entries if (ref $entries ne 'ARRAY');
+
+    write_sut_file(variable_file, join("\n", @$entries));
+    assert_script_run('source ' . variable_file, quiet => 1);
+}
+
+=head2 sdaf_get_deployer_ip
+
+    sdaf_get_deployer_ip(deployer_resource_group=>$deployer_resource_group);
+
+B<deployer_resource_group>: Deployer key vault name
+
+Retrieves public IP of the deployer VM.
+
+=cut
+
+sub sdaf_get_deployer_ip {
+    my (%args) = @_;
+    croak 'Missing "deployer_resource_group" argument' unless $args{deployer_resource_group};
+
+    my $vm_name = script_output("az vm list --resource-group $args{deployer_resource_group} --query [].name --output tsv");
+    my $az_query_cmd = join(' ', 'az', 'vm', 'list-ip-addresses', '--resource-group', $args{deployer_resource_group},
+        '--name', $vm_name, '--query', '"[].virtualMachine.network.publicIpAddresses[0].ipAddress"', '-o', 'tsv');
+
+    my $ip_addr = script_output($az_query_cmd);
+    croak "Not a valid ip addr: $ip_addr" unless grep /^$RE{net}{IPv4}$/, $ip_addr;
+    record_info('Deployer data', "Deployer resource group: $args{deployer_resource_group} \nDeployer VM IP: $ip_addr");
+    return $ip_addr;
+}
+
+=head2 sdaf_prepare_ssh_keys
+
+    sdaf_prepare_ssh_keys(deployer_key_vault=>$deployer_key_vault);
+
+B<deployer_key_vault>: Deployer key vault name
+
+Retrieves public and private ssh key from DEPLOYER keyvault and sets up permissions.
+
+=cut
+
+sub sdaf_prepare_ssh_keys {
+    my (%args) = @_;
+    croak 'Missing mandatory argument $args{deployer_key_vault}' unless $args{deployer_key_vault};
+    my $home = homedir();
+    my $az_cmd_out = script_output(
+        "az keyvault secret list --vault-name $args{deployer_key_vault} --query [].name --output tsv | grep ssh");
+    my %ssh_keys = (
+        id_rsa => grep(/sshkey$/, $az_cmd_out),
+        'id_rsa.pub' => grep(/sshkey-pub$/, $az_cmd_out)
```

Comment:
This will not work. `grep` expects a list as the second arg. But great idea! Suggestion:
```
    my @az_cmd_out = split(/\n/, script_output(
        "az keyvault secret list --vault-name $args{deployer_key_vault} --query [].name --output tsv | grep ssh"));
    my %ssh_keys = (
        id_rsa => grep(/sshkey$/, @az_cmd_out),
        'id_rsa.pub' => grep(/sshkey-pub$/, @az_cmd_out)
    );
```
I think you can even leave the `| grep ssh` out.

Reply:
Hmm.. this one does not seem to work. I think it might be that grep returns an array with one value. But this is working fine though:

```
    my %ssh_keys;
    my $az_cmd_out = script_output(
        "az keyvault secret list --vault-name $args{deployer_key_vault} --query [].name --output tsv | grep sshkey");

    foreach (split("\n", $az_cmd_out)) {
        $ssh_keys{id_rsa} = $_ if grep(/sshkey$/, $_);
        $ssh_keys{'id_rsa.pub'} = $_ if grep(/sshkey-pub$/, $_);
    }
```